src/lib/server/session.cpp

```diff
@@ -0,0 +1,200 @@
+#include "session.hpp"
+
+#include "postgres_message_types.hpp"
+#include "query_handler.hpp"
+#include "result_serializer.hpp"
+
+namespace opossum {
+
+Session::Session(boost::asio::io_service& io_service, const SendExecutionInfo send_execution_info)
+    : _socket(std::make_shared<Socket>(io_service)),
+      _postgres_protocol_handler(std::make_shared<PostgresProtocolHandler<Socket>>(_socket)),
+      _send_execution_info(send_execution_info) {}
+
+std::shared_ptr<Socket> Session::get_socket() { return _socket; }
+
+void Session::run() {
+  // Set TCP_NODELAY in order to disable Nagle's algorithm. It handles congestion control in TCP networks. Therefore,
+  // small packets are buffered and sent out later as one large packet. This might introduce a delay of up to 40 ms
+  // which we have to avoid. Further reading: https://howdoesinternetwork.com/2015/nagles-algorithm
+  _socket->set_option(boost::asio::ip::tcp::no_delay(true));
+  _establish_connection();
+  while (!_terminate_session) {
+    try {
+      _handle_request();
+    } catch (const std::exception& e) {
+      _postgres_protocol_handler->send_error_message(e.what());
+      _postgres_protocol_handler->send_ready_for_query();
+      // In order to avoid multiple ReadyForQuery in case of an error, we set this flag for further operations. As soon
+      // as a new query arrives it must be set to false again.
+      _sync_send_after_error = true;
```

Comment:
As a stand-alone comment, this is hard to understand. Any chance to add a sentence to make in clearer?